cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -38,8 +42,10 @@
   private static final long serialVersionUID = -1173500180640174909L;
 
   private Map<String, Map<String, Long>> mapReduceCounters;
-  private final Map<Scope, Map<String, List<NodeValue>>> tokenValueMap = new EnumMap<>(Scope.class);
+  private Map<Scope, Map<String, List<NodeValue>>> tokenValueMap =
+    Collections.synchronizedMap(new EnumMap<Scope, Map<String, List<NodeValue>>>(Scope.class));
   private String nodeName;
+  private List<String> nodeFilterList;
```

Comment:
Yes. Updated to the `set`. However its not final because the same instance of the `BasicWorkflowToken` is passed by driver to the different nodes before the execution (except fork). WorkflowDriver keep calling the `setCurrentNodeInfo` to update the `nodeName` and the `nodeFilterSet`.